### DISCUSSION TITLE
Simpler way to achieve labelled bar graph

### DISCUSSION THREAD

**Discussion by mcp292** [vega/altair#2988] [Q&A] [answered]
Is this still the simplest way to achieve a labelled bar graph?

https://altair-viz.github.io/gallery/bar_chart_with_labels.html

Is there a way to do this without layering? Possibly an encoding I am not aware of?

**Comment by joelostblom** [comment, marked as answer]
You will need to layer two marks to have both bars and text in the same chart. Having that said, the syntax of the example you linked could be more succinct:

```py
bars = alt.Chart(source).mark_bar().encode(
    x='wheat',
    y="year:O",
    text='wheat'
)

bars + bars.mark_text(align='left', dx=1)
```

This looks pretty readable to me, but I'm curious to hear what you had in mind for the syntax and if you have encountered issues with the layering approach.

**Reply by mcp292** [reply]
This is great!

As far as syntax I was hoping to not have to layer, but you've made sense of that. I was hoping to do it all in one graph, like with the `text` argument given, and not having to layer afterwards.

What you presented is indeed pretty succinct. I am surprised that `mark_bar().encode()` even provides a `text` parameter, based on your explanation.

**Reply by joelostblom** [reply]
> I am surprised that mark_bar().encode() even provides a text parameter, based on your explanation.

Technically it is the `Chart` object that takes the encoding, and then any number of marks can be added to it. The conventional order in the docs happens to be `alt.Chart().mark_*.encode()` but these methods could be added in any order and I could see an argument that it would be even more clear to have the encoding first:

```py
bars = alt.Chart(source).encode(
    x='wheat',
    y="year:O",
    text='wheat'
).mark_bar()

bars + bars.mark_text(align='left', dx=1)
```

or even:

```py
base = alt.Chart(source).encode(
    x='wheat',
    y="year:O",
    text='wheat'
)

base.mark_bar() + base.mark_text(align='left', dx=1)
```

Maybe one should even be allowed to chain marks in the same spec, but that is not supported currently and I am not sure how technically feasible that would be. And if you wanted to change an encoding for one of  the marks you would still need to layer as in the above, so probably not wortwhile to introduce mark chaining (it is different in ggplot where you can chain geoms since each geom also takes an optional `aes` parameter, but in Altair each mark does not take an optional `encode` parameter).

**Reply by mattijn** [reply]
> As far as syntax I was hoping to not have to layer

This is still work-in-progress at the Vega-Lite repository: https://github.com/vega/vega-lite/pull/7378.
That would mean that there eventually will be an encoding channel for `label`'s. This will avoids usage of a layered text-mark.

**Reply by mcp292** [reply]
> The conventional order in the docs happens to be alt.Chart().mark_*.encode() but these methods could be added in any order and I could see an argument that it would be even more clear to have the encoding first:

Wow, that does make it much more clear. 

> Maybe one should even be allowed to chain marks in the same spec

This is a cool idea!

**Reply by mcp292** [reply]
>  there eventually will be an encoding channel for `label`

That's going to be very convenient!

**Reply by mcp292** [reply]
I can see why `mark_*()` comes first though in the docs. You usually start by deciding what type of chart you want to plot, and then define the encoding accordingly. 

What might satisfy this scenario best, is using a base chart with just `encode()` and then doing bars and text off that, but that deviates from succinct.

**Reply by mattijn** [reply]
@joelostblom, I like this approach:

```python

base = alt.Chart(source).encode(
    x='wheat',
    y="year:O",
    text='wheat'
)

base.mark_bar() + base.mark_text(align='left', dx=1)
```
I think this style is better than the current approach in this example.

**Reply by mcp292** [reply]
This is exactly what I meant! 

I take it back this is very succinct. I must've been adding `mark` properties in my head.

**Reply by joelostblom** [reply]
> I think this style is better than the current approach in this example.

Yes, I like it better too, we should update that example. @mcp292 feel free to submit a PR if you would be interested in contributing since your inquiry sparked this discussion, otherwise I can do it later.

@mattijn what are your thoughts on using `encode()` before `mark()` in general? I quite like it because it focuses the thinking on the data first, and the "graphics" second. I am not sure if it is worthwhile changing it in the docs (it might be), but would be interested to hear your thoughts in general and if you think it would be beneficial for how to think about charts (I am leaning towards that it would be, even though marks are required while encodings are not). It is also similar to how things are conventionally written in ggplot with `ggplot() + aes() + geom_*()`, although those can also be remix in any order. The new [seaborn syntax also uses this order](https://seaborn.pydata.org/tutorial/objects_interface.html), but just because others do it doesn't mean it is the correct thing to do of course.

@mcp292, I think the point you brought up about thinking on the type of chart first is interesting and when I assess my own order of thinking, I would say that for myself I think about the columns that I want to plot first, and then which plot I would like to do. E.g. in the example above, my first thought would be something like "I want to see how the **wheat** yield changes over **time**" and then I would think of which mark type to use: bar, points etc. The opposite order would be odd for me: "I want to make a bar chart" and then "Which columns should I put in the barchart?". Curious to hear what you think.

**Reply by mcp292** [reply]
I appreciate it but you guys go ahead and take the credit on the PR. I would love to contribute but I'm pretty swamped with work at the moment, which is how all these questions are arising.

> for myself I think about the columns that I want to plot first, and then which plot I would like to do. E.g. in the example above, my first thought would be something like "I want to see how the wheat yield changes over time" and then I would think of which mark type to use: bar, points etc.

This is a great point actually, I guess I do think of the data encoding first and then the visualization follows naturally after I have established that. The one time this was different was yesterday when I plotted a horizontal bar graph, and the quantity I wanted to observe was on the `x` channel.

You know, I think I know why I said that, I usually map everything out on paper first, so by the time I go to code, I know I want a bar chart, and then figure what goes where. So the difference between us might be where that thinking happens. Yours might happen in the editor.

**Reply by joelostblom** [reply]
OK PR here https://github.com/altair-viz/altair/pull/2991

> I usually map everything out on paper first

I think this is a great practice and can be very helpful! I do it sometimes myself, mainly when planning for charts and datasets that I am less familiar with, or to get an overview early on in a project. Or when I get stuck on some detail of a visualization and need to take a step back. Definitely helps a lot, and I do start writing out the columns/encodings on the xy-axis on the paper before adding the marks, so my process is similar here to (but as you said, when transitioning to coding from paper it might differ)

**Reply by mattijn** [reply]
I can understand you from a theoretical point of view, but my practical experience differs a bit. I think I start from the mark. Like, I need a map that shows x, or I need a roseplot that shows y.

And then I often start with a verbose approach, especially for a layered charts, so that I can visualize each sublayer individually. 
Only in the last part I start merging and beautify things together, like what is suggested here.

**Reply by joelostblom** [reply]
Thanks for sharing that @mattijn. I got more curious and tried to look back in the history of the repo a bit. It seems like the syntax actually used to be `Chart().encode().mark_*()` in the early days, so there is probably some good reason that it changed. I couldn't find the specific PR that changed it, but this one include changing some of the examples https://github.com/altair-viz/altair/pull/107/files and this one clearly shows the old syntax https://github.com/altair-viz/altair/pull/92. Maybe I will look into this more in the future, but I'll leave it here for now.

**Reply by mcp292** [reply]
> It seems like the syntax actually used to be Chart().encode().mark_*() in the early days, so there is probably some good reason that it changed.

Fascinating! Nice digging!!
